### ISSUE TITLE
Question: ServletContainerInitializer from war file *no longer* called after adding <ordering> to web-fragment.xml from external jar.

### ISSUE THREAD

**Issue by jvanheesch** [eclipse/jetty.project#3146]
I extracted the issue in a small self-contained application (jetty-maven-plugin, version 9.4.14.v20181114):
https://github.com/jvanheesch/servlet-container-initializer-web-fragment-bug

The code includes:
- 1 webapp module (war), with a barebones web.xml, that depends on 2 jars (defined in the same multi-module project): external-jar and another-external-jar.
- The war and 2 jar artifacts all define an implementation of ServletContainerInitializer, declared in resources/META-INF/services/javax.servlet.ServletContainerInitializer.
- another-external-jar contains a web-fragment.xml, which I copied from log4j-web-2.10.0.jar.

The issue:
- If web-fragment.xml from another-external-jar declares no `<ordering>` element, then all three ServletContainerInitializer are called on container startup (mvn jetty:run from webapp directory).
- If the `<ordering>` element gets added to this web-fragment.xml, then SomeServletContainerInitializer.onStartup() is no longer called on container startup.

I noticed the same behavior in tomcat 9.0.12, but since the behavior surprised me, I was wondering if this behavior is either correct, incorrect, undefined, or something else.

**Comment by joakime** [comment]
Including text here for others that are searching for similar answers in the future.

The relevant text of the `web-fragment.xml`, and what is causing the problems OP has reported, is ...

``` xml
<web-fragment xmlns="http://java.sun.com/xml/ns/javaee"
              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              xsi:schemaLocation="http://java.sun.com/xml/ns/javaee
                                  http://java.sun.com/xml/ns/javaee/web-fragment_3_0.xsd"
              version="3.0" metadata-complete="true">
    <name>another-external-jar-logging</name>
    <distributable />
    <ordering>
        <before>
            <others />
        </before>
    </ordering>
</web-fragment>
```

**Comment by janbartel** [comment]
@jvanheesch thanks for the repro web app. Makes things much easier.

Firstly, let me say upfront that I think this is a bug on our behalf, so I've filed PR #3158  for it.  I can clearly see comments in the code where it looks like we were expecting to handle the case where the SCI comes from inside the webapp itself, however it looks like we were testing for this case in the wrong way  - the o.e.j.u.Resource representing the SCI class will not be null.

Having said that, I think it really is a grey area in the spec how/if that's supposed to work at all. If you read the spec section 8.2 Pluggability (p. 95 for Servlet Spec 4.0), it only ever mentions SCIs in relation to jars. The possibility of the SCI being in WEB-INF/classes is never addressed at all.  So I think jetty is (intending to be, and with the PR will be) lenient on this. You may want to raise an issue with the servlet spec to really get a proper clarification of this.

**Comment by janbartel** [comment]
Fix applied in time for jetty-9.4.15
